src/gil.rs

```diff
@@ -20,6 +20,9 @@ thread_local! {
     /// pub(crate) because it is manipulated temporarily by Python::allow_threads
     pub(crate) static GIL_COUNT: Cell<u32> = Cell::new(0);
 
+    /// Counter purely for GILGuards, to ensure they drop in the same order they are created.
+    static GILGUARD_COUNT: Cell<usize> = Cell::new(0);
```

Comment:
`bool` is sufficient. I'll explain why in the below comment.

Reply:
I don't think it is. Consider the case where three GILGuards get created:

```
let gil1 = Python::acquire_gil();
let gil2 = Python::acquire_gil();
let gil3 = Python::acquire_gil();

drop(gil2);
```

What should dropping `gil2` do? It can't modify the `bool`, because we're still in a nested case with `gil3`.

Reply:
Ah yes you're right, please don't mind the word